test/TestAppUtils/VisualTreeDumper.cs

```diff
@@ -0,0 +1,229 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Text;
+using Windows.UI.Xaml;
+using Windows.UI.Xaml.Media;
+
+namespace MUXControls.TestAppUtils
+{
+    class VisualTreeDumper
+    {
+        interface IVisitor
+        {
+            bool ShouldVisitNode(DependencyObject node);
+            void BeginVisitNode(DependencyObject node);
+            void EndVisitNode(DependencyObject node);
+
+            bool ShouldVisitPropertiesForNode(DependencyObject node);
+            bool ShouldVisitProperty(PropertyInfo propertyInfo);
+            void VisitProperty(String propertyName, Object value);
+        }
+
+        class Visitor : IVisitor
+        {
+            private StringBuilder _sb;
+            private int _indent;
+
+            public Visitor()
+            {
+                _sb = new StringBuilder();
+                _sb.AppendLine("<?xml version=\"1.0\" encoding=\"utf-8\" ?>");
+                _indent = 0;
+            }
+            public void EndVisitNode(DependencyObject obj)
+            {
+                _indent--;
+                AddPadding(_indent);
+                _sb.AppendLine("</Element>");
+            }
+
+            public void BeginVisitNode(DependencyObject obj)
+            {
+                AddPadding(_indent);
+                _sb.AppendLine(String.Format("<Element Type=\"{0}\">", obj.GetType().FullName));
+                _indent++;
+            }
+
+            public override String ToString()
+            {
+                return _sb.ToString();
+            }
+
+            public bool ShouldVisitNode(DependencyObject node)
+            {
+                return node != null;
+            }
+
+            public bool ShouldVisitPropertiesForNode(DependencyObject node)
+            {
+                return (node as UIElement) != null;
+            }
+
+            public bool ShouldVisitProperty(PropertyInfo propertyInfo)
+            {
+                var name = propertyInfo.Name;
+                return (_propertyNameBlackList.Where(item => name.EndsWith(item)).Count()) == 0 &&
+                    !_propertyNameBlackList.Contains(name);
+            }
+            public void VisitProperty(string propertyName, object value)
+            {
+                var v = PropertyValueToString(propertyName, value);
+                if (!IsKnownPropertyValue(propertyName, v))
```

Comment:
Why do we skip those "knowPropertyValues"? Is the tree too big so we wanted to skip uninteresting property values?

Reply:
yes. It's used to reduce the output. I was trying to setup a DefaultValue set, but that's not public API, so I did a KnownPropertyValues.

---
In reply to: [243777264](https://github.com/Microsoft/microsoft-ui-xaml/pull/126#discussion_r243777264) [](ancestors = 243777264)

Reply:
Default has special meaning for dependencyobjects, and those API is not public and we can't get it.

---
In reply to: [245380892](https://github.com/Microsoft/microsoft-ui-xaml/pull/126#discussion_r245380892) [](ancestors = 245380892)